File: LogicAndTrick.WikiCodeParser/Tags/ImageTag.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using LogicAndTrick.WikiCodeParser.Nodes;

namespace LogicAndTrick.WikiCodeParser.Tags
{
    public class ImageTag : Tag
    {
        public ImageTag()
        {
            Token = "img";
            Element = "span";
            MainOption = "url";
            Options = new[] {"url"};
        }

        public override INode FormatResult(Parser parser, ParseData data, State state, string scope, Dictionary<string, string> options, string text)
        {
            var url = text;
            if (options.ContainsKey("url")) url = options["url"];
            if (!Regex.IsMatch(url, "^([a-z]{2,10}://)", RegexOptions.IgnoreCase)) url = "http://" + url;
            url = System.Web.HttpUtility.HtmlAttributeEncode(url);

            var classes = new List<string>{"embedded", "image"};
            if (ElementClass != null) classes.Add(ElementClass);
            if (Token == "simg") classes.Add("inline");

            var element = Element;
            if (!classes.Contains("inline"))
            {
                state.SkipWhitespace();
                element = "div";
            }

            var before = $"<{element} class=\"{String.Join(" ", classes)}\">" +
                         $"<span class=\"caption-panel\">" +
                         $"<img class=\"caption-body\" src=\"{url}\" alt=\"User posted image\" />";
            var after = $"</span></{element}>";
            var plainsp = element == "div" ? "\n" : "";
            return new HtmlNode(before, PlainTextNode.Empty, after)
            {
                PlainBefore = $"{plainsp}[User posted image]{plainsp}"
            };
        }

        public override bool Validate(Dictionary<string, string> options, string text)
        {
            var url = text;
            if (options.ContainsKey("url")) url = options["url"];
            return !url.Contains("<script") && Regex.IsMatch(url, @"^([a-z]{2,10}://)?([^]""\n ]+?)$", RegexOptions.IgnoreCase);
        }
    }
}

File: php/src/Tags/ImageTag.php
<?php

namespace LogicAndTrick\WikiCodeParser\Tags;

use LogicAndTrick\WikiCodeParser\HtmlHelper;
use LogicAndTrick\WikiCodeParser\Nodes\HtmlNode;
use LogicAndTrick\WikiCodeParser\Nodes\INode;
use LogicAndTrick\WikiCodeParser\Nodes\PlainTextNode;
use LogicAndTrick\WikiCodeParser\ParseData;
use LogicAndTrick\WikiCodeParser\Parser;
use LogicAndTrick\WikiCodeParser\State;

class ImageTag extends Tag
{
    public function __construct()
    {
        parent::__construct();
        $this->token = 'img';
        $this->element = 'div';
        $this->mainOption = 'url';
        $this->options = ['url'];
        $this->isBlock = true;
    }

    public function FormatResult(Parser $parser, ParseData $data, State $state, string $scope, array $options, string $text): INode
    {
        $url = $text;
        if (isset($options['url'])) $url = $options['url'];
        if (!preg_match('/^([a-z]{2,10}:\/\/)/i', $url)) $url = 'http://' . $url;
        $url = HtmlHelper::AttributeEncode($url);

        $classes = ['embedded', 'image'];
        if ($this->elementClass != null) $classes[] = $this->elementClass;
        $element = $this->element;

        if (!$this->isBlock) {
            $element = 'span';
            $classes[] = 'inline';
        } else {
            $state->SkipWhitespace();
        }

        $before = '<' . $element . ' class="' . implode(' ', $classes) . '">' .
            '<span class="caption-panel">' .
            '<img class="caption-body" src="' . $url . '" alt="User posted image" />';
        $after = "</span></$element>";
        $plainsp = $element == 'div' ? "\n" : '';
        $ret = new HtmlNode($before, PlainTextNode::Empty(), $after);
        $ret->plainBefore = $plainsp . "[User posted image]" . $plainsp;
        $ret->isBlockNode = $element == 'div';
        return $ret;
    }

    public function Validate(array $options, string $text): bool
    {
        $url = $text;
        if (isset($options['url'])) $url = $options['url'];
        return !str_contains($url, '<script') && preg_match('/^([a-z]{2,10}:\/\/)?([^\]"\n ]+?)$/i', $url);
    }
}